On batch-msg dispatching: broker should disconnect consumer which doesn't support batch-message

rdhabalia opened the pull request:
### Motivation

Broker should verify consumer's version before dispatching `batch-messages` to the consumer. Right now, broker dispatches the batch-messages to consumer regardless consumer support it and it causes failure while client-consumer consumes the batch-message if consumer doesn't support batch-message feature.

### Modifications

Disconnect the consumer while dispatching `batch-message` if consumer doesn't support batch-message.

### Result

Broker will not dispatch batch-message if consumer doesn't support it.

rdhabalia commented:
>The only issue I see is that, the consumer will keep reconnecting in a loop without backoff.
The backoff time in reconnection is reset every time a producer/consumer is successful in re-establishing itself to the broker.

That is correct. But we are targeting clients which are using old version so, we have to prevent at broker side. To address reconnection, I modified change:
- close consumer at broker side without closing connection
- this will remove consumer from subscription and `ServerCnx` and resends pendingAck messages to other consumers

This will prevent reconnection, logs message at broker and consumer will not receive any further messages.

Any thought?

merlimat commented:
Sound good to me. Another option I was thinking of would to mark the topic as a "topic that a some point got batched messages" and prevent consumer to subscribe again. That way the back off would kick in

rdhabalia commented:
>Another option I was thinking of would to mark the topic as a "topic that a some point got batched messages" and prevent consumer to subscribe again. That way the back off would kick in

Yes, that is also a good option and at least in that way, client will receive an error and consumer would know what's going on.

rdhabalia commented:
@merlimat I think I misunderstood your previous comment. I have updated the change. Can you please take a look of it. I have introduced new error-code [UnsupporteVersion](https://github.com/rdhabalia/pulsar/blob/71fd6e2556d3c29e641e41751d11a9c766794597/pulsar-common/src/main/proto/PulsarApi.proto#L80) in case if one introduces new client without inital support of batch-msg, can also handle the error gracefully.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -213,9 +226,10 @@ int getBatchSizeforEntry(ByteBuf metadataAndPayload) {
+    int updatePermitsAndPendingAcks(final List<Entry> entries) throws PulsarServerException {
         int permitsToReduce = 0;
         Iterator<Entry> iter = entries.iterator();
+        boolean unsupportedVersion = false;
```
What about : 

```java
boolean clientSupportBatchMessages = cnx.isBatchMessageCompatibleVersion();
```

merlimat commented on pulsar-common/src/main/proto/PulsarApi.proto:
```diff
@@ -77,6 +77,7 @@ enum ServerError {
     ProducerBlockedQuotaExceededError = 7; // Unable to create producer because backlog quota exceeded
     ProducerBlockedQuotaExceededException = 8; // Exception while creating producer because quota exceeded
     ChecksumError = 9; // Error while verifying message checksum
+    UnsupportedVersionError = 10; // Error when broker/client doesn't support requested feature 
```
.. when an older client/version doesn't support a required feature

rdhabalia commented:
updated the changes.
